File: Data Structures & Algorithms/count-paths/submission-2.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        grid = [[-1] * n for _ in range(m)]

        def dfs(i: int, j: int) -> int:
            if i >= m or j >= n:
                return 0

            if grid[i][j] != -1:
                return grid[i][j]

            if i == m - 1 and j == n - 1:
                return 1


            grid[i][j] = dfs(i + 1, j) + dfs(i, j + 1)
            return grid[i][j]
    
        return dfs(0, 0)

File: Data Structures & Algorithms/island-perimeter/submission-2.cpp
class Solution {
public:
    int perim = 0;

    void dfs(int i, int j,
             vector<vector<int>>& grid,
             vector<vector<bool>>& visited) {

        if (i < 0 || j < 0 ||
            i >= grid.size() || j >= grid[0].size()) {
            perim++;
            return;
        }

        if (grid[i][j] == 0) {
            perim++;
            return;
        }

        if (visited[i][j]) {
            return;
        }

        visited[i][j] = true;

        dfs(i + 1, j, grid, visited);
        dfs(i - 1, j, grid, visited);
        dfs(i, j + 1, grid, visited);
        dfs(i, j - 1, grid, visited);
    }

    int islandPerimeter(vector<vector<int>>& grid) {
        vector<vector<bool>> visited(
            grid.size(),
            vector<bool>(grid[0].size(), false));

        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (grid[i][j] == 1) {
                    dfs(i, j, grid, visited);
                    return perim;
                }
            }
        }

        return 0;
    }
};